### PR TITLE
Fix issue #19300

### PR DESCRIPTION
**Pull request by raghavanone**
# What does this PR do?

This PR fixing issue #19300 

Fixes # (issue)
#19300

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@sgugger

### CODE REVIEW

**Comment by raghavanone** [comment]
@sgugger  How do I get output_dir in on_train_end callback ?

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1648,7 +1648,20 @@ def _inner_training_loop(
...
+                from ray import tune
+
+                run_id = tune.get_trial_id()
+            elif self.hp_search_backend == HPSearchBackend.SIGOPT:
+                run_id = trial.id
+            elif self.hp_search_backend == HPSearchBackend.WANDB:
+                import wandb
+
+                run_id = wandb.run.id
+        self.state.trial_name = self.hp_name(trial) if self.hp_name is not None else f"run-{run_id}"
```
Comment (sgugger) [inline src/transformers/trainer.py:1664]: I'm not sure why this is included in this PR, but it has nothing to do with the early stopping problem.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2097,6 +2110,9 @@ def _load_rng_state(self, checkpoint):
         if is_torch_tpu_available():
             xm.set_rng_state(checkpoint_rng_state["xla"])
 
+    def get_run_dir(self):
+        return os.path.join(self.args.output_dir, self.state.trial_name)
+
```
Comment (sgugger) [inline src/transformers/trainer.py:2115]: Same here.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -2118,8 +2134,7 @@ def _save_checkpoint(self, model, trial, metrics=None):
                 import wandb
 
                 run_id = wandb.run.id
-            run_name = self.hp_name(trial) if self.hp_name is not None else f"run-{run_id}"
-            run_dir = os.path.join(self.args.output_dir, run_name)
+            run_dir = self.get_run_dir()
```
Comment (sgugger) [inline src/transformers/trainer.py:2137]: And here.

File: src/transformers/trainer_callback.py [hunk 4]
```diff
@@ -563,6 +564,20 @@ def on_train_begin(self, args, state, control, **kwargs):
...
+        if (
+            state.best_model_checkpoint is not None
+            and state.args.save_total_limit == 1
+            and checkpoints_sorted[-1] != self.state.best_model_checkpoint
+        ):
+            checkpoints_to_be_deleted = checkpoints_sorted[-1]
+            for checkpoint in checkpoints_to_be_deleted:
+                logger.info(f"Deleting older checkpoint [{checkpoint}] due to args.save_total_limit")
+                shutil.rmtree(checkpoint)
+
```
Comment (sgugger) [inline src/transformers/trainer_callback.py:580]: The fix should be more general I think, and at the end of the `train` method itself.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. It includes some changes in the hyperparameter seach code which have nothing to do with the issue I believe, thos should be reverted.

As for the fix per se, I think we should have it at the end of the `train` method itself, not just for early stopping.

**Review by raghavanone — COMMENTED** [review]
@sgugger Some of the code is duplicated, should I pull it into separate method ?

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -1875,6 +1875,34 @@ def _inner_training_loop(
...
+            elif self.hp_search_backend == HPSearchBackend.SIGOPT:
+                run_id = trial.id
+            elif self.hp_search_backend == HPSearchBackend.WANDB:
+                import wandb
+
+                run_id = wandb.run.id
+            run_name = self.hp_name(trial) if self.hp_name is not None else f"run-{run_id}"
+            run_dir = os.path.join(self.args.output_dir, run_name)
+        else:
+            run_dir = self.args.output_dir
```
Comment (sgugger) [inline src/transformers/trainer.py:1896]: Yes, this whole block could probably be refactored in a method `_get_output_dir(self, trial)`

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -1875,6 +1875,34 @@ def _inner_training_loop(
...
+
+        output_dir = os.path.join(run_dir, checkpoint_folder)
+        checkpoints_sorted = self._sorted_checkpoints(use_mtime=True, output_dir=output_dir)
+
+        if self.state.best_model_checkpoint is not None and self.state.args.save_total_limit == 1:
```
Comment (sgugger) [inline src/transformers/trainer.py:1901]:
```suggestion
        # Delete the last checkpoint when save_total_limit=1 if it's different from the best checkpoint.
        if self.state.best_model_checkpoint is not None and self.state.args.save_total_limit == 1:
```

**Review by sgugger — COMMENTED** [review]
Thanks a lot for iterating on this! I think we can refactor the dupe code indeed.

File: src/transformers/trainer.py (hunk 5 above)
Comment (raghavanone) [reply src/transformers/trainer.py:1913]: Done , But at #L2158 there is "self.store_flos()" inside the else part, not show how to handle it .

Comment (sgugger) [reply src/transformers/trainer.py:1913]: You should:
1. use the `_get_output_dir` method
2. then make a test for `if self.hp_search_backend is None or trial is None:` and in it do the store_flos part.

**Review by sgugger — APPROVED** [review]
Very clean, thanks a lot for iterating!

Not sure why the tests are failing. Will have a deeper look later today.

**Comment by raghavanone** [comment]
@sgugger  Never mind, Let me fix the failing tests.

**Comment by sgugger** [comment]
Let me know if you need any help!

**Comment by raghavanone** [comment]
@sgugger  The tests pass now. There was bug in my change and it is great that our tests caught it.

**Comment by sgugger** [comment]
Thanks again for all your work on this!

**Pull request merged**